File: src/content/docs/topics/cs/coding-problems/linked-list/146-lru-cache-approach3.py
class Node:
    __slots__ = ("key", "val", "prev", "next")
    def __init__(self, key=0, val=0):
        self.key = key; self.val = val
        self.prev = None; self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}
        self.head = Node(); self.tail = Node()
        self.head.next = self.tail; self.tail.prev = self.head

    def _remove(self, node):
        node.prev.next = node.next; node.next.prev = node.prev

    def _add_to_front(self, node):
        node.prev = self.head; node.next = self.head.next
        self.head.next.prev = node; self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove(node); self._add_to_front(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]; node.val = value
            self._remove(node); self._add_to_front(node); return
        if len(self.cache) >= self.cap:
            lru = self.tail.prev
            self._remove(lru); del self.cache[lru.key]
        node = Node(key, value)
        self.cache[key] = node; self._add_to_front(node)

cache = LRUCache(2)
cache.put(1, 1); cache.put(2, 2)
assert cache.get(1) == 1
cache.put(3, 3)
assert cache.get(2) == -1
cache.put(4, 4)
assert cache.get(1) == -1
assert cache.get(3) == 3
assert cache.get(4) == 4
c1 = LRUCache(1)
c1.put(1, 10); assert c1.get(1) == 10
c1.put(2, 20); assert c1.get(1) == -1; assert c1.get(2) == 20
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/linked-list/146-lru-cache-approach3.swift
// LEETCODE_TYPE: LRUCache
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

private final class CacheNode {
    let key: Int
    var value: Int
    var previous: CacheNode?
    var next: CacheNode?

    init(_ key: Int, _ value: Int) {
        self.key = key
        self.value = value
    }
}

final class LRUCache {
    private let capacity: Int
    private var nodes: [Int: CacheNode] = [:]
    private let leastRecent = CacheNode(0, 0)
    private let mostRecent = CacheNode(0, 0)

    init(_ capacity: Int) {
        self.capacity = capacity
        leastRecent.next = mostRecent
        mostRecent.previous = leastRecent
    }

    func get(_ key: Int) -> Int {
        guard let node = nodes[key] else { return -1 }
        remove(node)
        insertMostRecent(node)
        return node.value
    }

    func put(_ key: Int, _ value: Int) {
        if let existing = nodes[key] {
            remove(existing)
            existing.value = value
            insertMostRecent(existing)
            return
        }
        let node = CacheNode(key, value)
        nodes[key] = node
        insertMostRecent(node)
        if nodes.count > capacity, let victim = leastRecent.next, victim !== mostRecent {
            remove(victim)
            nodes[victim.key] = nil
        }
    }

    private func remove(_ node: CacheNode) {
        node.previous?.next = node.next
        node.next?.previous = node.previous
    }

    private func insertMostRecent(_ node: CacheNode) {
        let previous = mostRecent.previous
        previous?.next = node
        node.previous = previous
        node.next = mostRecent
        mostRecent.previous = node
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:f3bf00711ce9a2a8d1225cc7c7e80f04af6ee3b8d7f2e35883fcb953fa11b270
    let subject1 = LRUCache(2)
    subject1.put(1, 1)
    subject1.put(2, 2)
    expectEqual(subject1.get(1), 1, "canonical-eviction[3]")
    subject1.put(3, 3)
    expectEqual(subject1.get(2), -1, "canonical-eviction[5]")
    subject1.put(4, 4)
    expectEqual(subject1.get(1), -1, "canonical-eviction[7]")
    expectEqual(subject1.get(3), 3, "canonical-eviction[8]")
    expectEqual(subject1.get(4), 4, "canonical-eviction[9]")
    let subject2 = LRUCache(2)
    subject2.put(1, 1)
    subject2.put(1, 9)
    expectEqual(subject2.get(1), 9, "update-existing-key[3]")
    let subject3 = LRUCache(2)
    subject3.put(1, 1)
    subject3.put(2, 2)
    expectEqual(subject3.get(1), 1, "get-refreshes-recency[3]")
    subject3.put(3, 3)
    expectEqual(subject3.get(2), -1, "get-refreshes-recency[5]")
    let subject4 = LRUCache(1)
    subject4.put(1, 1)
    subject4.put(2, 2)
    expectEqual(subject4.get(1), -1, "capacity-one[3]")
    expectEqual(subject4.get(2), 2, "capacity-one[4]")
    // EXCLUDED_VECTOR zero-capacity: [[{"operation":"init","arguments":[0]}]] | Capacity must be at least one.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
